rest_framework/fields.py

```diff
@@ -532,6 +534,7 @@ def run_validators(self, value):
         errors = []
         for validator in self.validators:
             if hasattr(validator, 'set_context'):
+                validator = copy.deepcopy(validator)
```

Comment:
If `set_context` means that we need a separate instance, perhaps we should be reconsidering the interface (e.g. `set_context` returns a new instance).

deepcopy is black magic, we should avoid it if we can.

Reply:
My first thought was to pass the instance to the validator when calling it, and just pass it to all functions instead of saving it on the instance (ie. as `self.instance`).  But not all validators accept a context and using inspect to decide if the validator accepts a context seemed like a bad idea.  Last, but not least, this would break all existing custom validators with `set_context`.

Changing the interface of `set_context` is also a breaking change for all such custom validators.

So maybe this is something to consider for a 4.0 release, but not for 3.8?  I'd love to see a fix for this in the 3.8 release.  (I've patched the one place in our code that threw the most exceptions, but not all Validators in all our microservices.)

Reply:
It seems that deepcopy is a good way to fix the issue for now (I do not have an opinion about whether deepcopy is good, bad or magic). Changing the interface or the meaning of ``set_context`` is potentially a breaking change, though it would be for a very good reason.

Seeing what this bug is causing (potentially inconsistencies in the database), I believe that a quick fix should be applied.

For a long term solution, it should be considered to NOT store ``self.instance`` in the validators ``set_context`` method, but to pass it to the ``validator(...)`` method. Maybe it is possible to do both for some time and log a notification with a deprecation warning.

Reply:
> pass it to the `validator(...)` method

Do you have any good solution how to decide whether to pass it or not, ie. a replacement for `hasattr(validator, 'set_context')`.  It's possible to use any django validator (from the core, from drf, third party or custom) here and most of them do not accept the instance.

I could only think of using [inspect](https://docs.python.org/3/library/inspect.html) (+ py2.7) or `try-except`.

Reply:
True, that means all those validators would have to be changed, I guess that's not feasible at all (and, afterall, it is definately not going to happen).

Reply:
Possible solution: 
- Refactor unique validators to inherit a base `InstanceValidator`
- Change the `InstanceValidator` call signature to expect an `instance` in addition to the value.
- Delete the `set_context` method handling.
- Handle regular validators first, then instance validators second.

Reply:
Thanks for the input.  I like it and will look into it.

Reply:
I've finally gotten around to tackle this.  I've opened a new PR: encode/django-rest-framework#6172